src/models/src/entity/atproto.rs

```diff
@@ -0,0 +1,368 @@
+        // return early if we got any error
+        let (session_manager, _) = data.atproto.callback(params).await.map_err(|error| {
+            error!(%error, "failed to complete authorization callback for atproto");
+
+            ErrorResponse::new(
+                ErrorResponseType::BadRequest,
+                "failed to complete authorization callback for atproto",
+            )
+        })?;
+
+        let agent = Agent::new(session_manager);
+
+        let (did, email, email_confirmed) = match agent.api.com.atproto.server.get_session().await {
+            Ok(Object {
+                data:
+                    get_session::OutputData {
+                        did,
+                        email,
+                        email_confirmed,
+                        ..
+                    },
+                ..
+            }) => (did, email, email_confirmed),
+            Err(error) => {
+                error!(%error, "failed to get session for atproto");
+
+                return Err(ErrorResponse::new(
+                    ErrorResponseType::Internal,
+                    "failed to get session for atproto",
+                ));
+            }
+        };
+
+        let link_cookie = ApiCookie::from_req(req, PROVIDER_LINK_COOKIE)
+            .and_then(|value| AuthProviderLinkCookie::try_from(value.as_str()).ok());
+
+        if email.is_none() {
+            todo!()
+        }
+
+        let claims_user_id = did.to_string();
+
+        let user_opt = match User::find_by_federation("atproto", &claims_user_id).await {
+            Ok(user) => {
+                debug!(
+                    "found already existing user by federation lookup: {:?}",
+                    user
+                );
+                Some(user)
+            }
+            Err(_) => {
+                debug!("did not find already existing user by federation lookup - making sure email does not exist");
+
+                if let Ok(mut user) = User::find_by_email(email.as_ref().unwrap().to_string()).await
+                {
+                    if let Some(link) = link_cookie.as_ref() {
+                        if link.provider_id != "atproto" {
+                            return Err(ErrorResponse::new(
+                                ErrorResponseType::BadRequest,
+                                "bad provider_id in link cookie".to_string(),
+                            ));
+                        }
+
+                        if link.user_id != user.id {
+                            return Err(ErrorResponse::new(
+                                ErrorResponseType::BadRequest,
+                                "bad user_id in link cookie".to_string(),
+                            ));
+                        }
+
+                        if link.user_email != user.email {
+                            return Err(ErrorResponse::new(
+                                ErrorResponseType::BadRequest,
+                                "Invalid E-Mail".to_string(),
```

Comment:
You don't need the `.to_string()` allocation for the error reponse message.

Reply:
I just copy-pasted most of this code from `AuthProviderIdClaims`.

Reply:
Oh okay, if this still has some `.to_string()`, they can all be removed. I did this optimization some time ago and am not sure, if I catched all of the updates.